src/main/java/InfrastructureManager/MatchMaking/ScoreBasedMM.java

```diff
@@ -0,0 +1,228 @@
+package InfrastructureManager.MatchMaking;
+
+import InfrastructureManager.EdgeClient;
+import InfrastructureManager.EdgeNode;
+import InfrastructureManager.EdgeClientHistory;
+import com.google.common.collect.Multimap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * Score-based match making algorithm.
+ * Score = Wd_D + Wc_C + Wn_N + Wh_H --- D = ping, C = res, N = network, H = history
+ * Weight: 1 - 10 - 10 - 10
+ *
+ * @author Zero
+ */
+public class ScoreBasedMM implements MatchMakingAlgorithm {
+
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    // TODO: this should be dynamic
+    private static final long ACCEPTABLE_PING = 300; //use this as parameter before put them in the calculation
+    private static final long RESOURCE_WEIGHT = 10;
+    private static final long NETWORK_WEIGHT = 10;
+    private static final long PING_WEIGHT = 1;
+    private static final long HISTORY_WEIGHT = 10;
+    private static final long QOS_THRESHOLD = 100000;
+
+    Multimap<String, HashMap<String, Long>> nodeHistory;
+    EdgeClientHistory clientHistory;
+    long deductScorePerSecond = 1;
+
+    /* big step
+    DONE: Score = Wd_D + Wc_C + Wn_N + Wh_H --- D = ping, C = res, N = network, H = history
+        Weight: 1 - 10 - 10 - 10
+    DONE: implement slots (resource and network), this should be an attribute in edge client and edge note.
+    TODO: History implementation, using SQL? Or keep throwing JSON object back and forward?
+    DONE: Need to make sure there won't be no duplicate in nodeList (Id and Address)
+    TODO: Weight should be dynamic later on
+    TODO: Implement time
+
+     */
+
+    /*
+    TODO: implement visibility: blind-inacc-accurate. This rather connected to "input" analyze,
+     take the input and somehow "blur" the info by introduce some deviant (multiply by 5-30%, so we get the estimation)
+
+     */
+
+    /* small step
+    DONE: match base on available resources
+    DONE: eliminate nodes that can't accommodate client (not enough resources, network etc)
+    DONE: compare between "acceptable" node using score
+     */
+    @Override
+    public EdgeNode match(EdgeClient thisClient, List<EdgeNode> nodeListInput, EdgeClientHistory thisClientHistory) {
+        logger.info("""
+                
+                ----------------------------MATCH MAKING - SCORE BASED--------------------
+                R_weight, N_weight, D_weight, H_weight, QOS_threshold
+                {}      , {}      , {}      , {}      , {}    """
+                , RESOURCE_WEIGHT, NETWORK_WEIGHT, PING_WEIGHT, HISTORY_WEIGHT, QOS_THRESHOLD);
+
+        //Initiating variable
+        List<EdgeNode> nodeList = new ArrayList<>(nodeListInput); //In case of multi threading
```

Comment:
To avoid race condition, I just clone that nodeListInput array list to the nodeList, so whatever happens with nodeListInput if we decided to have several instances of matchmaking and SBMM that may modify that list, the list for each SBMM will not be affected.

E.g. we have 100 nodes, sent to SBMM for match making, but the moment we sent the list, 10 nodes get disconnected.
Now the question here is do we use the list that has 90 nodes, or 100 nodes? Or what if we update the nodes in that list. Should we use the original one or the modified one? 

I leave the comment say multi-threading mostly for those cases if I have to decide between either situation.

Reply:
The match making should snap a stable version of the list, it doesn't matter if the list changes afterwards, so what you are doing here is correct, you just need to add mutex here as well.

Reply:
Not sure how mutex should be added in this part tbh. 
We going to clone the input `nodeListInput`, then the mutex should be put in MatchMaker, not here (since we initiate and changing the list over there)

 Gonna leave it here. @alyhasansakr 
Reopen this convo if you think otherwise